text/0013-prepare-blockbuilder-and-core-runtime-apis-for-mbms.md

```diff
@@ -0,0 +1,120 @@
+# RFC-0013: Prepare `BlockBuilder` and `Core` runtime API for MBMs
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | July 24, 2023 |
+| **Description** | Prepares the `BlockBuilder` and `Core` Runtime APIs for Multi-Block-Migrations. |
+| **Authors**     | Oliver Tale-Yazdi |
+
+## Summary
+
+Introduces breaking changes to the `BlockBuilder` and `Core` runtime APIs and bumps them to version
+5 and 7 respectively. A new hook `after_inherents` is added to the `BlockBuilder` runtime API and
+the `initialize_block` function of `Core` is changed to return an enum to indicate what extrinsics
+and hooks can be executed.
+
+## Motivation
+
+The original motivation is Multi-Block-Migrations. They require the runtime to be able to tell the
+block builder that it should not attempt to include transactions in the current block. Currently,
+there is no communication possible between runtime and block builder that could achieve this.
+Further, it is necessary to execute logic right after inherent application but still before
+transaction inclusion.
+
+This RFC proposes a way of communication between the runtime and the block builder by changing the
+`initialize_block` function to return a `BlockExecutiveMode` enum. Additionally, an
+`after_inherents` function is introduced that runs after inherent but before transaction
+application.  
```

Comment:
Random remark: why not make `after_inherents` return the `BlockExecutiveMode`, instead of `initialize_block`? Then it would just indicate to the client whether the block authorship should continue or stop now.

Reply:
Thanks for the comment, originally it was implemented like this.  
But then came up the idea of optional inherents: https://github.com/paritytech/substrate/pull/14414#issuecomment-1606262686, and with that the need to already communicate this `Mode` before the inherents get applied. In general this allows for more freedom already before generating inherents. 

Does it sounds good to you?

Reply:
> then came up the idea of optional inherents

The comment you linked is Basti asking what optional inherents are, because as far as I understand there's no such thing as an optional inherent at the moment.

I'm not very opinionated, it just seems to me that the design is a tiny bit weirder than it could be.

Reply:
Yes the runtime could alter the return value of `inherent_extrinsics` :thinking:   

There is still this one [comment from Gav](https://github.com/paritytech/substrate/pull/14414#issuecomment-1602222428) which mentions to already having two modes prior to inherent application.

Reply:
`after_inherents` was now gladly removed.